drivers/net/mlx5/mlx5_vdpa.c

```diff
@@ -163,6 +200,122 @@ static int mlx5_vdpa_release_rx(struct vdpa_priv *priv)
+		(((mkey_attr->size + 4095) / 4096) + 1) / 2);
+	MLX5_SET(create_mkey_in, in, translations_octword_actual_size,
+		(((mkey_attr->size + 4095) / 4096) + 1) / 2);
+	MLX5_SET(create_mkey_in, in, pg_access, 1);
+	MLX5_SET64(mkc, mkc, start_addr, mkey_attr->addr);
+	MLX5_SET64(mkc, mkc, len, mkey_attr->size);
+	MLX5_SET(mkc, mkc, log_page_size, 12);
+	MLX5_SET(create_mkey_in, in, mkey_umem_id, mkey_attr->pas_id);
+
+	mkey->obj = mlx5_glue->dv_devx_obj_create(ctx, in, sizeof(in), out,
+					   sizeof(out));
+	if (!mkey->obj) {
+		DRV_LOG(ERR, "Can't create mkey error %s\n", strerror(errno));
+		goto error;
+	}
+	status = MLX5_GET(create_mkey_out, out, status);
+	mkey->key = MLX5_GET(create_mkey_out, out, mkey_index);
+	mkey->key = (mkey->key << 8) | 0x50;
+	DRV_LOG(DEBUG, "create mkey status %d mkey value %d\n",
+		status, (mkey->key));
+	if (status)
+		goto error;
+	return mkey;
+error:
+	if (mkey)
+		rte_free(mkey);
+	return NULL;
+}
+
+static
+struct mlx5_devx_mkey *mlx5_create_indirect_mkey(struct ibv_context *ctx,
+					struct mlx5_devx_mkey_attr *mkey_attr,
+					struct mlx5_klm *klm_array, int num_klm)
+{
+	int translations_oct_size = (((num_klm / 4) + (num_klm % 4)) * 4);
+	uint32_t in_size = MLX5_ST_SZ_DB(create_mkey_in) +
+				translations_oct_size * MLX5_ST_SZ_DB(klm);
+	uint32_t *in = rte_zmalloc("in", in_size, 64);
+	uint32_t out[MLX5_ST_SZ_DW(create_mkey_out)] = {0};
+	uint32_t status;
+	void *mkc;
+	struct mlx5_devx_mkey *mkey = NULL;
+	uint8_t *klm = (uint8_t *)MLX5_ADDR_OF(create_mkey_in, in, klm_pas_mtt);
+
+	mkey = rte_zmalloc("mkey", sizeof(*mkey), 64);
+	MLX5_SET(create_mkey_in, in, opcode, MLX5_CMD_OP_CREATE_MKEY);
+	mkc = MLX5_ADDR_OF(create_mkey_in, in, memory_key_mkey_entry);
+	MLX5_SET(mkc, mkc, lw, 0x1);
+	MLX5_SET(mkc, mkc, lr, 0x1);
+	MLX5_SET(mkc, mkc, rw, 0x1);
+	MLX5_SET(mkc, mkc, rr, 0x1);
+	MLX5_SET(mkc, mkc, access_mode_1_0, MLX5_MKC_ACCESS_MODE_KSM);
+	MLX5_SET(mkc, mkc, qpn, 0xffffff);
+	MLX5_SET(mkc, mkc, length64, 0x0);
+	MLX5_SET(mkc, mkc, pd, mkey_attr->pd);
+	MLX5_SET(mkc, mkc, mkey_7_0, 0x50);//FIXME: should be dynamic
+	MLX5_SET(mkc, mkc, translations_octword_size, translations_oct_size);
+	MLX5_SET(create_mkey_in, in,
+		translations_octword_actual_size, translations_oct_size);
+	MLX5_SET(create_mkey_in, in, pg_access, 0);
+	MLX5_SET64(mkc, mkc, start_addr, mkey_attr->addr);
+	MLX5_SET64(mkc, mkc, len, mkey_attr->size);
+	MLX5_SET(mkc, mkc, log_page_size, max(mkey_attr->log_entity_size, 12));
+	for (int i = 0; i < num_klm; i++) {
+		MLX5_SET(klm, klm, mkey, klm_array[i].mkey);
+		MLX5_SET64(klm, klm, address, klm_array[i].address);
+		klm += MLX5_ST_SZ_DB(klm);
+	}
+	for (int i = num_klm; i < translations_oct_size; i++) {
+		MLX5_SET(klm, klm, mkey, 0x0);
+		MLX5_SET64(klm, klm, address, 0x0);
+		klm += MLX5_ST_SZ_DB(klm);
+	}
+	mkey->obj = mlx5_glue->dv_devx_obj_create(ctx, in, in_size, out,
+						  sizeof(out));
+	if (!mkey->obj) {
+		DRV_LOG(ERR, "Can't create mkey error %s\n", strerror(errno));
```

Comment:
if possible - remove the \n (later when running on DEBUG it creates an redundant empty line)